src/main/scala/io/citrine/lolo/trees/splits/BoltzmannSplitter.scala

```diff
@@ -0,0 +1,226 @@
+  * from the mean).  This is akin to kinetic monte carlo and simulated annealing techniques.
+  *
+  * Created by maxhutch on 11/29/16.
+  *
+  * @param temperature used to control how sensitive the probability of a split is to its change in variance
+  */
+case class BoltzmannSplitter(temperature: Double) extends Splitter[Double] {
+
+  /**
+    * Get the a split probabalisticly, considering numFeature random features (w/o replacement), ensuring that the
+    * resulting partitions have at least minInstances in them
+    *
+    * @param data         to split
+    * @param numFeatures  to consider, randomly
+    * @param minInstances minimum instances permitted in a post-split partition
+    * @return a split object that optimally divides data
+    */
+  def getBestSplit(data: Seq[(Vector[AnyVal], Double, Double)], numFeatures: Int, minInstances: Int): (Split, Double) = {
+    /* Pre-compute these for the variance calculation */
+    val calculator = VarianceCalculator.build(data.map(_._2), data.map(_._3))
+    val initialVariance = calculator.getImpurity
+    val beta = 1.0 / (temperature * initialVariance)
+
+    val rep = data.head
+
+    /* Try every feature index */
+    val featureIndices: Seq[Int] = rep._1.indices
+
+    val possibleSplits: Seq[SplitterResult] = Random.shuffle(featureIndices).take(numFeatures).flatMap { index =>
+      /* Use different spliters for each type */
+      rep._1(index) match {
+        case _: Double => BoltzmannSplitter.getBestRealSplit(data, calculator, index, minInstances, beta)
+        case _: Char => BoltzmannSplitter.getBestCategoricalSplit(data, calculator, index, minInstances, beta)
+        case _: Any => throw new IllegalArgumentException("Trying to split unknown feature type")
+      }
+    }
+
+    // If we couldn't find a split, then return NoSplit with no variance reduction
+    if (possibleSplits.isEmpty) {
+      return (new NoSplit(), 0.0)
+    }
+
+    // Re-based the probabilities, such that the largest probability is order-1.0
+    // This is meant to avoid every probability underflowing
+    val rebase = possibleSplits.map(_.base).max
+    val totalProbability = possibleSplits.map { x => x.rebasedScore * Math.exp(x.base - rebase) }.sum
+
+    // select from a discrete probability distribution by drawing a random number and then computing the CDF
+    // where the "draw" is the bin for which the CDF crosses the drawn number
+    val draw = Random.nextDouble() * totalProbability
+    // could be a scanLeft + find, but this is more readable
+    var cumSum: Double = 0.0
+    possibleSplits.foreach { case SplitterResult(split, variance, score, base) =>
+      // Here's the probability rebasing again
+      cumSum = cumSum + score * Math.exp(base - rebase)
+      if (draw < cumSum) {
+        val deltaImpurity = initialVariance - variance
+        return (split, deltaImpurity)
+      }
```

Comment:
Since this logic is repeated in both boltzmann splitters, might want to pull it out as a utility function